### PR TITLE
Execute a process with a compensation end event

### LINKED ISSUE #14967: I can execute a process with a compensation end event
## Description 

As a user, I can deploy a process with a compensation end event. Now, I want to execute the process. In the first step, we want to ignore the invocation of compensation handlers. 

- I can complete a process instance that contains a compensation end event.
- The compensation end event is activated and completed.
- The records for the compensation end event have the event type `COMPENSATION`.

Screenshot from 2023-11-02 14-01-15

requires #14942 
requires #14944

### PR DESCRIPTION
**Pull request by nicpuppa**
## Description

With this PR is now possible to execute and complete a process with a Compensation End event.

**NOTE**

Compensation handler are not supported yet

## Related issues

closes #14967 

⚠️ ⚠️ **WARNING** ⚠️ ⚠️ 

This PR depends on [15361](https://github.com/camunda/zeebe/pull/15361)

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/EndEventProcessor.java [hunk 1]
```diff
@@ -313,4 +314,29 @@ public void onComplete(final ExecutableEndEvent element, final BpmnElementContex
...
+    }
+
+    @Override
+    public void onComplete(final ExecutableEndEvent element, final BpmnElementContext completing) {
+      stateTransitionBehavior
+          .transitionToCompleted(element, completing)
+          .ifRightOrLeft(
+              stateTransitionBehavior::completeElement,
+              failure -> incidentBehavior.createIncident(failure, completing));
+    }
```
Comment (saig0) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/EndEventProcessor.java:340]: :x: We're making an invalid transition. The end event is in the state `COMPLETED` and we write a `COMPLETE` command. 

In the record stream, you can see that the `COMPLETE` command is rejected. 

```
 - #26->#20 K10 - SEQUENCE_FLOW "sequenc..1b104eb" in <process "compens..process"[K03]>
C PI        ACTIVATE       - #27->#20 K11 - END_EVENT "compens..n-event" in <process "compens..process"[K03]>
E PI        ACTIVATING     - #28->#20 K11 - END_EVENT "compens..n-event" in <process "compens..process"[K03]>
E PI        ACTIVATED      - #29->#20 K11 - END_EVENT "compens..n-event" in <process "compens..process"[K03]>
E PI        COMPLETING     - #30->#20 K11 - END_EVENT "compens..n-event" in <process "compens..process"[K03]>
E PI        COMPLETED      - #31->#20 K11 - END_EVENT "compens..n-event" in <process "compens..process"[K03]>
C PI        COMPLETE       - #32->#20 K03 - PROCESS "compens..process" in <process "compens..process"[K03]>
C PI        COMPLETE       - #33->#20 K11 - END_EVENT "compens..n-event" in <process "compens..process"[K03]>
E PI        COMPLETING     - #34->#20 K03 - PROCESS "compens..process" in <process "compens..process"[K03]>
E PI        COMPLETED      - #35->#20 K03 - PROCESS "compens..process" in <process "compens..process"[K03]>
R PI        COMPLETE       - #36->#20 K11 - END_EVENT "compens..n-event" in <process "compens..process"[K03]> !INVALID_STATE (Expected element instance with key '2251799813685259' to be present in state but not found.)
```

I recommend that we follow the other behaviors and call `stateTransitionBehavior.takeOutgoingSequenceFlows(element, completed)`.

**Review by saig0 — CHANGES_REQUESTED** [review]
@nicpuppa looks good but I discovered one issue. :grin: Please have a look. :cookie:

File: engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/EndEventProcessor.java (hunk 1 above)
Comment (nicpuppa) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/EndEventProcessor.java:340]: Thanks for spotting 👀

**Review by saig0 — APPROVED** [review]
@nicpuppa LGTM. :rocket:

**Comment by nicpuppa** [comment]
bors merge

**Pull request merged**